feat(docs): using telepresence in development mode. (#567)
kubewarden/kubewarden-controller: CONTRIBUTING.md

# Contributing

> This document contains instructions on how to build and run locally the controller

Reviewer: ```suggestion
This document contains instructions on how to build and run the controller locally
```

  enabled. If not provided and `WEBHOOK_HOST_LISTEN` is provided,
  it will be defaulted to `WEBHOOK_HOST_LISTEN`.

The Subject Alternative Names of the generated certificate in development mode
will contain whatever was provided on `WEBHOOK_HOST_ADVERTISE` (or whatever it
was defaulted to, if it was not provided).

To run the controller in the developer workstation it's possible to use
[telepresence](https://github.com/telepresenceio/telepresence). It allows
developers to "add" the local workstation in the cluster. Thus, it's possible
to access workloads in the cluster and receive requests to the controller running
in the developer machine.  

> > Install the telepresence binary from the Github releases page. The official 
> > once from Ambassador Labs may require login. 

Reviewer: ```suggestion
> Install the telepresence binary from the Github releases page. The official 
> ones from Ambassador Labs may require login. 
```

Before running the controller, install the custom resource definitions:

```console
kubectl apply -f config/crd/bases
```

Create the `kubewarden` namespace

```console
kubectl create ns kubewarden
```

Now, as we are using `telepresence` to intercept the intra cluster
communication, we need to create the controller deployment and service that
`telepresence` will intercept. For this, install the Kuberwarden stack

> following the steps described quickstart guide. In other words, install

Reviewer: ```suggestion
following the steps described in the quickstart guide. In other words, install
```

> Kubewarden controller helm chart. However, a change is required. The controller

Reviewer: ```suggestion
the Kubewarden controller helm chart. However, a change is required. The controller
```

> deployment must allow root users. This is disable by default and it is required
> because `telepresence` will add a init container in the deployment which need
> root access.

Reviewer: ```suggestion
deployment must allow root users. This is disabled by default and it is required
because `telepresence` will add a init container in the deployment which needs
root access.
```

```

In order to avoid two controllers trying to reconcile the same resources, it's necessary to
remove the deployment installed by the `kubewarden-controller` helm chart and deploy a dummy
deployment just to allow `telepresence` to intercept the communications.

```console
kubectl delete deployment -n kubewarden kubewarden-controller
kubectl apply -f - <<EOF
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app.kubernetes.io/instance: kubewarden-controller
    app.kubernetes.io/name: kubewarden-controller
  name: kubewarden-controller
  namespace: kubewarden
spec:
  selector:
    matchLabels:
      app.kubernetes.io/instance: kubewarden-controller
      app.kubernetes.io/name: kubewarden-controller
  template:
    metadata:
      labels:
        app.kubernetes.io/instance: kubewarden-controller
        app.kubernetes.io/name: kubewarden-controller
    spec:
      containers:
      - name: echo-server
        image: gcr.io/google-containers/echoserver:1.8
        ports:
        - containerPort: 9443
      serviceAccount: kubewarden-controller
      serviceAccountName: kubewarden-controller
  replicas: 1
EOF
```

> > Note: if you want to use some kind o NetworkPolicy to block the original controller traffic
> > you may need to have a multi node cluster. Because in a single node cluster the traffic 
> > policies will not be applied. Because the traffic will not leave the node and never the
> > checked against the network policies

Reviewer: ```suggestion
> Note: if you want to use some kind of NetworkPolicy to block the original controller traffic
> you may need to have a multi node cluster. In a single node cluster the traffic 
> policies will not be applied. Because the traffic will not leave the node and will never be
> checked against the network policies
```
